exla/lib/exla/builder.ex

```diff
@@ -3,25 +3,75 @@ defmodule EXLA.Builder do
   Wrapper around XLA's builder.
   """
 
-  alias __MODULE__
-  alias EXLA.{Computation, Op}
+  alias EXLA.Computation
+  alias EXLA.Op
+  alias EXLA.MLIR.Module, as: M
+  alias EXLA.MLIR.Type
 
   @enforce_keys [:ref]
   defstruct [:ref, :parent, :name]
 
+  def new(name, _inputs, _outputs, :xla) do
+    new(name)
+  end
+
+  def new(name, inputs, outputs, :mlir) do
+    mlir_arg_types = mlir_type(Enum.map(inputs, &elem(&1, 1)))
+    mlir_ret_type = mlir_type(outputs)
+
+    xla_ret_shape = exla_shape(outputs)
+
+    module = M.new()
+    M.create_function(module, "main", mlir_arg_types, mlir_ret_type, xla_ret_shape)
+  end
+
+  defp mlir_type(input) when is_tuple(input) do
+    input
+    |> Tuple.to_list()
+    |> Enum.map(&mlir_type/1)
+  end
+
+  defp mlir_type(inputs) when is_list(inputs) do
+    Enum.map(inputs, &mlir_type/1)
+  end
+
+  defp mlir_type(%EXLA.Shape{} = shape) do
+    Type.new(shape)
+  end
+
+  defp mlir_type(%Nx.Tensor{} = t) do
+    t.type
+    |> EXLA.Shape.make_shape(t.shape)
+    |> Type.new()
+  end
+
+  defp exla_shape(%Nx.Tensor{} = t) do
+    EXLA.Shape.make_shape(t.type, t.shape)
+  end
+
   def new(name) when is_binary(name) do
     {:ok, ref} = EXLA.NIF.new_builder(name)
-    %Builder{ref: ref, parent: nil, name: name}
+    %__MODULE__{ref: ref, parent: nil, name: name}
   end
 
-  def new(builder = %Builder{ref: ref}, name) when is_binary(name) do
+  def new(builder = %__MODULE__{ref: ref}, name) when is_binary(name) do
     {:ok, ref} = EXLA.NIF.create_sub_builder(ref, name)
-    %Builder{ref: ref, parent: builder, name: name}
+    %__MODULE__{ref: ref, parent: builder, name: name}
   end
```

Comment:
Btw, this is used in nested scopes such as `if`, `cond`. So I imagine we will add another branch in the future. :)

Reply:
Yes, I only implemented the minimal changes to get things to work. The new branch would have been untested, so I didn't bother